### ISSUE TITLE
Backport statedb backend system from sharding branch to master

### ISSUE THREAD

**Issue by pipermerriam** [ethereum/py-evm#216]
### What should be done

#213 implements a backend system for the `StateDB` class to facilitate protocol changes to how account state is stored.  This API is currently only present in the `sharding` branch (once #213 is merged).

The `TwoLayerBackend` and the changes to the `StateDB` class should be merged into master separately from the `sharding` branch.

**Comment by pipermerriam** [comment]
This is superceded by #507
